File: ios/App/Repository/CloudRepo.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2021 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Combine

class CloudRepo {

    // Whether DNS profile is currently selected or not, refreshed on foreground
    var dnsProfileActivatedHot: AnyPublisher<CloudDnsProfileActivated, Never> {
        self.writeDnsProfileActivated.compactMap { $0 }.removeDuplicates().eraseToAnyPublisher()
    }

    // The latest device info from our backend, including blocklists, device tag, etc
    var deviceInfoHot: AnyPublisher<DevicePayload, Never> {
        self.writeDeviceInfo.compactMap { $0 }.eraseToAnyPublisher()
    }

    // Lastest device tag info from backend. Used for Cloud DNS profile and Plus VPN
    var deviceTagHot: AnyPublisher<CloudDeviceTag, Never> {
        self.deviceInfoHot.map { it in it.device_tag }.removeDuplicates().eraseToAnyPublisher()
    }

    // Latest user blocklists choice (only backend IDs, needs to be translated client side)
    var blocklistsHot: AnyPublisher<CloudBlocklists, Never> {
        self.deviceInfoHot.map { it in it.lists }.removeDuplicates().eraseToAnyPublisher()
    }
    
    // User selected activity retention (may be empty for "no retention")
    var activityRetentionHot: AnyPublisher<CloudActivityRetention, Never> {
        self.deviceInfoHot.map { it in it.retention }.removeDuplicates().eraseToAnyPublisher()
    }

    // Whether user has currently paused adblocking or not
    var adblockingPausedHot: AnyPublisher<Bool, Never> {
        self.deviceInfoHot.map { it in it.paused }.removeDuplicates().eraseToAnyPublisher()
    }

    private lazy var api = Services.apiForCurrentUser
    private lazy var privateDns = Services.privateDns

    private lazy var envRepo = Repos.envRepo
    private lazy var processingRepo = Repos.processingRepo
    private lazy var enteredForegroundHot = Repos.stageRepo.enteredForegroundHot
    private lazy var activeTabHot = Repos.navRepo.activeTabHot
    private lazy var accountIdHot = Repos.accountRepo.accountIdHot

    private let bgQueue = DispatchQueue(label: "CloudRepoBgQueue")

    fileprivate let writeDnsProfileActivated = CurrentValueSubject<CloudDnsProfileActivated?, Never>(nil)
    fileprivate let writeDeviceInfo = CurrentValueSubject<DevicePayload?, Never>(nil)

    fileprivate let setActivityRetentionT = Tasker<CloudActivityRetention, Ignored>("setActivityRetention")
    fileprivate let refreshDeviceInfoT = SimpleTasker<Ignored>("refreshDeviceInfo", debounce: 3.0, errorIsMajor: true)
    fileprivate let setPausedT = Tasker<Ignored, Ignored>("setPaused", errorIsMajor: true)

    private var cancellables = Set<AnyCancellable>()

    init() {
        onRefreshDeviceInfo()
        onSetActivityRetention()
        onSetPaused()
        onForegroundCheckDnsProfileActivation()
        onTabChangeRefreshDeviceInfo()
        onDeviceTagChangeUpdateDnsProfile()
        onAccountIdChangeRefreshDeviceInfo()
    }

    func setActivityRetention(_ retention: CloudActivityRetention) -> AnyPublisher<Ignored, Error> {
        return setActivityRetentionT.send(retention)
    }

    func setPaused(_ paused: Bool) -> AnyPublisher<Ignored, Error> {
        return self.setPausedT.send(paused)
    }

    private func onRefreshDeviceInfo() {
        refreshDeviceInfoT.setTask { _ in Just(true)
            .flatMap { _ in self.api.getDeviceForCurrentUser() }
            .tryMap { it in
                self.writeDeviceInfo.send(it)
                return true
            }
            .eraseToAnyPublisher()
        }
    }
    
    private func onSetActivityRetention() {
        setActivityRetentionT.setTask { retention in Just(retention)
            .flatMap { it in self.api.putActivityRetentionForCurrentUser(it) }
            .tryMap { it in
                self.refreshDeviceInfoT.send()
                return true
            }
            .eraseToAnyPublisher()
        }
    }

    private func onSetPaused() {
        setPausedT.setTask { paused in Just(paused)
            .flatMap { it in self.api.putPausedForCurrentUser(it) }
            .tryMap { it in
                self.refreshDeviceInfoT.send()
                return true
            }
            .eraseToAnyPublisher()
        }
    }

    // Will check the activation status on every foreground event
    private func onForegroundCheckDnsProfileActivation() {
        enteredForegroundHot
        .flatMap { _ in self.privateDns.isPrivateDnsProfileActive() }
        .sink(
            onValue: { isActivated in self.writeDnsProfileActivated.send(isActivated) }
        )
        .store(in: &cancellables)
    }

    // Will recheck device info on each tab change.
    // This struct contains something important for each tab.
    // Entering foreground will also re-publish active tab even if user doesn't change it.
    private func onTabChangeRefreshDeviceInfo() {
        activeTabHot
        .sink(onValue: { it in self.refreshDeviceInfoT.send() })
        .store(in: &cancellables)
    }

    // Whenever device tag from backend changes, update the DNS profile in system settings.
    // User is meant to activate it manually, but our app can update it anytime.
    private func onDeviceTagChangeUpdateDnsProfile() {
        deviceTagHot
        .flatMap { it in self.privateDns.savePrivateDnsProfile(tag: it, name: self.envRepo.deviceName) }
        .sink(
            onFailure: { err in self.processingRepo.notify(self, err, major: true) }
        )
        .store(in: &cancellables)
    }

    // Whenever account ID is changed, device tag will change, among other things.
    func onAccountIdChangeRefreshDeviceInfo() {
        accountIdHot
        .sink(onValue: { it in self.refreshDeviceInfoT.send() })
        .store(in: &cancellables)
    }

}

class DebugCloudRepo: CloudRepo {

    private let log = Logger("Cloud")
    private var cancellables = Set<AnyCancellable>()

    override init() {
        super.init()

        dnsProfileActivatedHot.sink(
            onValue: { it in self.log.v("dnsProfileActivated: \(it)") }
        )
        .store(in: &cancellables)

        deviceInfoHot.sink(
            onValue: { it in self.log.v("deviceInfo: \(it)") }
        )
        .store(in: &cancellables)

    }
}

File: android5/app/src/main/java/repository/CloudRepo.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2022 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package repository

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import model.CloudActivityRetention
import model.CloudBlocklists
import model.DevicePayload
import model.Granted
import service.BlockaApiForCurrentUserService
import service.ConnectivityService
import service.EnvironmentService
import service.Services
import utils.Ignored
import utils.Logger
import utils.SimpleTasker
import utils.Tasker

open class CloudRepo {

    private val env by lazy { EnvironmentService }
    private val api by lazy { Services.apiForCurrentUser }
    private val connectivity by lazy { ConnectivityService }

    private val enteredForegroundHot by lazy { Repos.stage.enteredForegroundHot }
    private val accountIdHot by lazy { Repos.account.accountIdHot }
    private val activeTabHot by lazy { Repos.nav.activeTabHot }

    private val writeDeviceInfo = MutableStateFlow<DevicePayload?>(null)
    private val writeDnsProfileActivated = MutableStateFlow<Granted?>(null)
    private val writePrivateDnsSetting = MutableStateFlow<String?>(null)

    val deviceInfoHot = writeDeviceInfo.filterNotNull().distinctUntilChanged()

    val expectedDnsStringHot = deviceInfoHot.map {
        // TODO: better sanitize device name
        val deviceName = env.getDeviceAlias().replace(" ", "--")
        val tag = it.device_tag
        "$deviceName-$tag.cloud.blokada.org"
    }

    val dnsProfileActivatedHot = writeDnsProfileActivated.filterNotNull().distinctUntilChanged()

    val deviceTagHot = deviceInfoHot.map { it.device_tag }.distinctUntilChanged()
    val blocklistsHot = deviceInfoHot.map { it.lists }.distinctUntilChanged()
    val activityRetentionHot = deviceInfoHot.map { it.retention }.distinctUntilChanged()
    val adblockingPausedHot = deviceInfoHot.map { it.paused }.distinctUntilChanged()

    private val refreshDeviceInfoT = SimpleTasker<Ignored>("refreshDeviceInfo", debounce = 500, errorIsMajor = true)
    private val setActivityRetentionT = Tasker<CloudActivityRetention, Ignored>("setActivityRetention")
    private val setBlocklistsT = Tasker<CloudBlocklists, Ignored>("setBlocklists")
    private val setPausedT = Tasker<Boolean, Ignored>("setPaused", errorIsMajor = true)

    open fun start() {
        GlobalScope.launch { onRefreshDeviceInfo() }
        GlobalScope.launch { onSetActivityRetention() }
        GlobalScope.launch { onSetBlocklists() }
        GlobalScope.launch { onSetPaused() }

        GlobalScope.launch { onTabChange_refreshDeviceInfo() }
        GlobalScope.launch { onAccountIdChanged_refreshDeviceInfo() }
        GlobalScope.launch { onPrivateDnsProfileChanged_update() }

        onPrivateDnsSettingChanged_update()
    }

    suspend fun setActivityRetention(retention: CloudActivityRetention) {
        setActivityRetentionT.send(retention)
    }

    suspend fun setPaused(paused: Boolean) {
        setPausedT.send(paused)
    }

    suspend fun setBlocklists(lists: CloudBlocklists) {
        setBlocklistsT.send(lists)
    }

    private suspend fun onRefreshDeviceInfo() {
        refreshDeviceInfoT.setTask {
            writeDeviceInfo.value = api.getDeviceForCurrentUser()
            true
        }
    }

    private suspend fun onSetActivityRetention() {
        setActivityRetentionT.setTask {
            api.putActivityRetentionForCurrentUser(it)
            refreshDeviceInfoT.send()
        }
    }

    private suspend fun onSetPaused() {
        setPausedT.setTask {
            api.putPausedForCurrentUser(it)
            refreshDeviceInfoT.send()
        }
    }

    private suspend fun onSetBlocklists() {
        setBlocklistsT.setTask {
            api.putBlocklistsForCurrentUser(it)
            refreshDeviceInfoT.send()
        }
    }

    // Will recheck device info on each tab change.
    // This struct contains something important for each tab.
    // Entering foreground will also re-publish active tab even if user doesn't change it.
    private suspend fun onTabChange_refreshDeviceInfo() {
        activeTabHot
        .collect {
            refreshDeviceInfoT.send()
        }
    }

    // Whenever account ID is changed, device tag will change, among other things.
    private suspend fun onAccountIdChanged_refreshDeviceInfo() {
        accountIdHot
        .collect {
            refreshDeviceInfoT.send()
        }
    }

    private suspend fun onPrivateDnsProfileChanged_update() {
        expectedDnsStringHot
        .combine(writePrivateDnsSetting) { setting, expected -> setting == expected }
        .collect { writeDnsProfileActivated.value = it }
    }

    private fun onPrivateDnsSettingChanged_update() {
        connectivity.onPrivateDnsChanged = {
            writePrivateDnsSetting.value = it
        }
        // First check has to happen manually
        writePrivateDnsSetting.value = connectivity.privateDns
    }

}

class DebugCloudRepo: CloudRepo() {

    override fun start() {
        super.start()

        GlobalScope.launch { printDnsProfAct() }
        GlobalScope.launch { printDeviceInfo() }
    }

    private suspend fun printDnsProfAct() {
        dnsProfileActivatedHot
        .collect {
            Logger.v("Cloud", "dnsProfileActivated: $it")
        }
    }

    private suspend fun printDeviceInfo() {
        deviceInfoHot
        .collect {
            Logger.v("Cloud", "deviceInfo: $it")
        }
    }

}